# Урок 10. Дисперсионный анализ

Импорт необходимых библиотек:

In [54]:
import numpy as np
from scipy import stats

## Задача №1

_Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов.
Даны значения роста в трех группах случайно выбранных спортсменов:
Футболисты: **173, 175, 180, 178, 177, 185, 183, 182.**
Хоккеисты: **177, 179, 180, 188, 177, 172, 171, 184, 180.**
Штангисты: **172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.**_

In [55]:
footballers_height = np.array([173, 175, 180, 178, 177, 185, 183, 182])
hockey_players_height = np.array([177, 179, 180, 188, 177, 172, 171, 184, 180])
weightlifters_height = np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170])

In [56]:
alpha = 0.05

#### Проверка данных на нормальность распределения

In [64]:
stat_fb, p_val_fb = stats.shapiro(footballers_height)
print(f'Футболисты: stat_fb = {stat_fb:.3f}, p_val_fb = {p_val_fb:.3f}')
stat_hp, p_val_hp = stats.shapiro(hockey_players_height)
print(f'Хоккеисты: stat_hp = {stat_hp:.3f}, p_val_hp = {p_val_hp:.3f}')
stat_wl, p_val_wl = stats.shapiro(weightlifters_height)
print(f'Штангисты: stat_wl = {stat_wl:.3f}, p_val_wl = {p_val_wl:.3f}')

Футболисты: stat_fb = 0.978, p_val_fb = 0.950
Хоккеисты: stat_hp = 0.958, p_val_hp = 0.776
Штангисты: stat_wl = 0.939, p_val_wl = 0.505


In [63]:
print(f'Футболисты: ({p_val_fb:.3f} > {alpha}) == {p_val_fb > alpha}\n'
      f'Хоккеисты: ({p_val_hp:.3f} > {alpha}) == {p_val_hp > alpha}\n'
      f'Штангисты: ({p_val_wl:.3f} > {alpha}) == {p_val_wl > alpha}')
if p_val_fb > alpha and p_val_hp > alpha and p_val_wl > alpha:
    print('pvalue для каждой выборки больше alpha. Данные выборок имеют нормальное распределение.')
else:
    print('pvalue одной или нескольких выборок меньше alpha. Данные выборок не имеют нормального распределения.')

Футболисты: (0.950 > 0.05) == True
Хоккеисты: (0.776 > 0.05) == True
Штангисты: (0.505 > 0.05) == True
pvalue для каждой выборки больше alpha. Данные выборок имеют нормальное распределение.


#### Проверка равенства дисперсий

In [59]:
stat_sportsmen, p_val_sportsmen = stats.bartlett(footballers_height, hockey_players_height, weightlifters_height)
print(f'Спортсмены: stat_sportsmen = {stat_sportsmen:.3f}, p_val_sportsmen = {p_val_sportsmen:.3f}')

Спортсмены: stat_sportsmen = 0.464, p_val_sportsmen = 0.793


In [65]:
print(f'Спортсмены: ({p_val_sportsmen:.3f} > {alpha}) == {p_val_sportsmen > alpha}')
if p_val_sportsmen > alpha:
    print('pvalue для каждой выборки больше alpha. Дисперсии равны.')
else:
    print('pvalue для каждой выборки меньше alpha. Дисперсии не равны.')

Спортсмены: (0.793 > 0.05) == True
pvalue для каждой выборки больше alpha. Дисперсии равны.


#### Однофакторный дисперсионный анализ

Формулировка гипотез:
$H0:\mu_{1}=\mu_{2}=\mu_{3}$    -    __Футболисты, хоккеисты и штангисты не имеют существенных различий в росте.__
$H1:\mu_{1}\neq\mu_{2}\neq\mu_{3}$    -    __Футболисты, хоккеисты и штангисты имеют существенные различия в росте.__

In [61]:
stat_ua, p_val_ua = stats.f_oneway(footballers_height, hockey_players_height, weightlifters_height)
print(f'Средний рост спортсменов: stat_ua = {stat_ua:.5f}, p_val_ua = {p_val_ua:.5f}')

Средний рост спортсменов: stat_ua = 5.50005, p_val_ua = 0.01048


Получили значение $pvalue = 0.01048$ на уровне статистической значимости $\alpha = 0.05$.

In [67]:
print(f'Средний рост спортсменов: ({p_val_ua:.5f} > {alpha}) == {p_val_ua > alpha}')
if p_val_ua > alpha:
    print('Футболисты, хоккеисты и штангисты не имеют существенных различий в росте.\nНулевая гипотеза принимается.')
else:
    print('Футболисты, хоккеисты и штангисты имеют существенные различия в росте.\nНулевая гипотеза отвергается.')

Средний рост спортсменов: (0.01048 > 0.05) == False
Футболисты, хоккеисты и штангисты имеют существенные различия в росте.
Нулевая гипотеза отвергается.
